docs/how-it-works.md

```diff
@@ -0,0 +1,59 @@
+# How it works
+
+The objective of this document explains how the NGINX Ingress controller works, in particular how the NGINX model is built and why we need a one.
+
+## NGINX configuration
+
+The goal of this Ingress controller is the assembly of a configuration file (nginx.conf). The main implication of this requirement is the need to reload NGINX after any change in the configuration file.
+
+## NGINX model
+
+Usually, a Kubernetes Controller utilizes the [synchronization loop pattern](1) to check if the desired state in the controller is updated or a change is required. To this purpose, we need to build a model using different objects from the cluster, in particular (in no special order) Ingresses, Services, Endpoints, Secrets, and Configmaps to generate a point in time configuration file that reflects the state of the cluster.
+
+To get this object from the cluster, we use [Kubernetes Informers](2), in particular, `FilteredSharedInformer`. This informers allows reacting to changes in using [callbacks](3) to individual changes when a new object is added, modified or removed. Unfortunately, there is no way to know if a particular change is going to affect the final configuration file. Therefore on every change, we have to rebuild a new model from scratch based on the state of cluster and compare it to the current model. If the new model equals to the current one, then we avoid generating a new NGINX configuration and [trigger a reload](7). Otherwise, we create a new NGINX configuration based on the new model, replace the current model and [trigger a reload](7).
+
+One of the uses of the model is to avoid unnecessary reloads when there's no change in the state and to detect conflicts in definitions.
+
+The final representation of the model is generated from a [Go template](6) and the model that reflects the NGINX configuration.
```

Comment:
should not this be final representation of Nginx configuration? I commented at https://github.com/kubernetes/ingress-nginx/pull/2479#discussion_r190894770